server/datastore/mysql/queries.go

```diff
@@ -523,24 +527,43 @@ func (ds *Datastore) ListQueries(ctx context.Context, opt fleet.ListQueryOptions
 		}
 	}
 
-	if opt.MatchQuery != "" {
-		whereClauses += " AND q.name = ?"
-		args = append(args, opt.MatchQuery)
+	// normalize the name for full Unicode support (Unicode equivalence).
+	normMatch := norm.NFC.String(opt.MatchQuery)
+	whereClauses, args = searchLike(whereClauses, args, normMatch, querySearchColumns...)
+
+	getQueriesStmt += whereClauses
+
+	// build the count statement before adding pagination constraints
+	getQueriesCountStmt := fmt.Sprintf("SELECT COUNT(DISTINCT id) FROM (%s) AS s", getQueriesStmt)
+
+	getQueriesStmt, args = appendListOptionsWithCursorToSQL(getQueriesStmt, args, &opt.ListOptions)
+
+	dbReader := ds.reader(ctx)
+	queries := []*fleet.Query{}
+	if err := sqlx.SelectContext(ctx, dbReader, &queries, getQueriesStmt, args...); err != nil {
+		return nil, 0, nil, ctxerr.Wrap(ctx, err, "listing queries")
 	}
 
-	sql += whereClauses
-	sql, args = appendListOptionsWithCursorToSQL(sql, args, &opt.ListOptions)
+	// perform a second query to grab the count
+	var count int
+	if err := sqlx.GetContext(ctx, dbReader, &count, getQueriesCountStmt, args...); err != nil {
+		return nil, 0, nil, ctxerr.Wrap(ctx, err, "get queries count")
+	}
 
-	results := []*fleet.Query{}
-	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &results, sql, args...); err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "listing queries")
+	if err := ds.loadPacksForQueries(ctx, queries); err != nil {
+		return nil, 0, nil, ctxerr.Wrap(ctx, err, "loading packs for queries")
 	}
 
-	if err := ds.loadPacksForQueries(ctx, results); err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "loading packs for queries")
+	var meta *fleet.PaginationMetadata
+	if opt.ListOptions.IncludeMetadata {
+		meta = &fleet.PaginationMetadata{HasPreviousResults: opt.ListOptions.Page > 0}
+		if len(queries) > int(opt.ListOptions.PerPage) { //nolint:gosec // dismiss G115
```

Comment:
Won't `len(queries)` always be <= `int(opt.ListOptions.PerPage)`?

Maybe you need to use `count`, `page` and `per_page`?

It's possible I'm confused.

Reply:
Ah, answering myself, `appendListOptionsWithCursorToSQL` is including one extra item when `opt.ListOptions.IncludeMetadata == true` (Let's add a comment somewhere for future readers? it's not obvious :).